subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/CrossProjectConfigurationReportingTaskExecutionGraph.kt

```diff
@@ -0,0 +1,245 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache
+
+import groovy.lang.Closure
+import org.gradle.api.Action
+import org.gradle.api.InvalidUserCodeException
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.execution.TaskExecutionGraph
+import org.gradle.api.execution.TaskExecutionGraphListener
+import org.gradle.api.execution.TaskExecutionListener
+import org.gradle.api.internal.project.CrossProjectModelAccess
+import org.gradle.api.internal.project.ProjectInternal
+import org.gradle.configuration.internal.UserCodeApplicationContext
+import org.gradle.configurationcache.extensions.capitalized
+import org.gradle.configurationcache.problems.ProblemsListener
+import org.gradle.configurationcache.problems.PropertyProblem
+import org.gradle.configurationcache.problems.StructuredMessage
+import org.gradle.configurationcache.problems.location
+import org.gradle.execution.plan.FinalizedExecutionPlan
+import org.gradle.execution.plan.Node
+import org.gradle.execution.taskgraph.TaskExecutionGraphInternal
+import org.gradle.internal.build.ExecutionResult
+import org.gradle.util.Path
+import java.util.Objects
+import java.util.function.Consumer
+
+
+class CrossProjectConfigurationReportingTaskExecutionGraph(
+    private val delegate: TaskExecutionGraphInternal,
+    private val referrerProject: ProjectInternal,
+    private val problems: ProblemsListener,
+    private val crossProjectModelAccess: CrossProjectModelAccess,
+    private val coupledProjectsListener: CoupledProjectsListener,
+    private val userCodeContext: UserCodeApplicationContext
+) : TaskExecutionGraphInternal {
+    override fun addTaskExecutionGraphListener(listener: TaskExecutionGraphListener) {
+        delegate.addTaskExecutionGraphListener(listener.wrap())
+    }
+
+    override fun removeTaskExecutionGraphListener(listener: TaskExecutionGraphListener) {
+        delegate.removeTaskExecutionGraphListener(listener.wrap())
+    }
+
+    override fun whenReady(closure: Closure<*>) {
+        delegate.whenReady(CrossProjectModelAccessTrackingClosure(closure, referrerProject, crossProjectModelAccess))
+    }
+
+    override fun whenReady(action: Action<TaskExecutionGraph>) {
+        delegate.whenReady(action.wrap())
+    }
+
+    override fun hasTask(path: String): Boolean {
+        // If the queried task is in the referrer project, it's basically not a case of cross-project configuration;
+        // If there are only tasks from the current project in the task graph, then it's safe, too.
+        val parentPath = Path.path(path).parent?.path
+        if (parentPath != referrerProject.path && delegate.allTasks.any { it.project.path != parentPath }) {
```

Comment:
It seems to me that even when there are only tasks from the current project in the task graph, an attempt to reach out to a task from a different project should be reported (and cause coupling) since that project might get reconfigured later and that might cause the queried task to appear in the execution graph.

I'd try to reuse the work that already happens for `hasTask` by extracting a `@Nullable Task findTask(String path)` and, with that in place, write something like:

```kotlin
    override fun findTask(path: String): Task? {
        return delegate.findTask(path).also { task ->
            if (task == null) {
                // check whether path refers to a different project
                val parentPath = Path.path(path).parent?.path
                if (parentPath != referrerProject.path) {
                    val coupledProjects = listOfNotNull(parentPath?.let { referrerProject.findProject(it) })
                    reportCrossProjectTaskAccess(coupledProjects, path)
                }
            } else {
                checkCrossProjectTaskAccess(task)
            }
        }
    }

    override fun hasTask(path: String): Boolean {
        return findTask(path) != null
    }
```

Reply:
Hmm, you are right about querying for tasks from another project even if they are not there!
I've done what you suggested in https://github.com/gradle/gradle/pull/22445/commits/bcf919d0b73afd6ce4f45d0feee4bf405d3a3224.
Thanks!

Reply:
❌ Hmm, there's probably one more tricky case of potential project coupling around the task graph. There might be not a single task from a project `foo` in the graph, but `foo` could still participate in task graph construction. For example, `foo` might have a `Configuration` with empty build dependencies which some other project chooses as a project dependency resolution result.
Now if the `foo`'s configuration outcome changes, the dependency might bring some `foo`'s tasks to the task graph. So whoever observed the task graph's `getAllTasks` or `getFilteredTasks` is effectively coupled with `foo`. 

Let me think how we can handle this. It might be possible to keep track of the projects visited while collecting the build dependencies. This sounds like a correct but complex approach. 

Could coupling with all the configured projects be OK too?

Reply:
This seems to also affect the other PR covering `TaskDependency.getDependencies(...)`. And in that case, coupling the violating project with every other configured project does not seem to be practical.

Reply:
I've filed #22564 for this problem.